filters/encrypt/filter.go

```diff
@@ -0,0 +1,542 @@
+	// for the FilterOperations to be applied to DataClassifications.
+	//
+	// Normally, the filter operation applied to a field is determined by the
+	// operation specified in it's "classified" tag. If no operation is
+	// specified in the tag, then a set of reasonable default filter operations
+	// are applied.
+	//
+	// FilterOperationOverrides provides the ability to override an event's
+	// "classified" tag settings.
+	FilterOperationOverrides map[DataClassification]FilterOperation
+
+	l sync.RWMutex
+}
+
+// Reopen is a no op for Filters.
+func (af *Filter) Reopen() error {
+	return nil
+}
+
+// Type describes the type of the node as a Filter.
+func (ef *Filter) Type() eventlogger.NodeType {
+	return eventlogger.NodeTypeFilter
+}
+
+// Rotate supports rotating the filter's wrapper, salt and info via the options:
+// WithWrapper, WithSalt, WithInfo
+func (ef *Filter) Rotate(opt ...Option) {
+	opts := getOpts(opt...)
+	ef.l.Lock()
+	defer ef.l.Unlock()
+	if opts.withWrapper != nil {
+		ef.Wrapper = opts.withWrapper
+	}
+	if opts.withSalt != nil {
+		ef.HmacSalt = opts.withSalt
+	}
+	if opts.withInfo != nil {
+		ef.HmacInfo = opts.withInfo
+	}
+}
+
+// Process will encrypt or hmac-sha256 string and []byte fields which are tagged
+// as SensitiveClassification.  Fields that are tagged SecretClassification will
+// be redacted.
+//
+// If the event payload satisfies the WrapperPayload interface, then the
+// payload's Wrapper(), HmacSalt() and HmacInfo() will be used to rotate the
+// filter's wrappers for ongoing filtering operations.  Events matching this
+// WrapperPayload interface are not sent along in the pipeline and a nil with no
+// errors is immediately returned after the wrapper has been rotated.
+//
+// If the event payload satisfies the EventWrapperInfo interface, then the
+// payload's EventId(), HmacSalt() and HmacInfo() will be used to for filtering
+// operations for just the single event being processed.
+func (ef *Filter) Process(ctx context.Context, e *eventlogger.Event) (*eventlogger.Event, error) {
+	const op = "event.(Filter).Process"
+	if e == nil {
+		return nil, fmt.Errorf("%s: missing event: %w", op, ErrInvalidParameter)
+	}
+	if i, ok := e.Payload.(RotateWrapper); ok {
+		ef.l.Lock()
+		defer ef.l.Unlock()
+		if i.Wrapper() != nil {
+			ef.Wrapper = i.Wrapper()
+		}
+		if i.HmacSalt() != nil {
+			ef.HmacSalt = make([]byte, len(i.HmacSalt()))
+			copy(ef.HmacSalt, i.HmacSalt())
+		}
+		if i.HmacInfo() != nil {
+			ef.HmacInfo = make([]byte, len(i.HmacInfo()))
+			copy(ef.HmacInfo, i.HmacInfo())
+		}
+		return nil, nil
+	}
+
+	var opts []Option
+	var optWrapper wrapping.Wrapper
+	if i, ok := e.Payload.(EventWrapperInfo); ok {
+		ef.l.RLock()
+		w, err := NewEventWrapper(ef.Wrapper, i.EventId())
+		ef.l.RUnlock()
+		if err != nil {
+			return nil, fmt.Errorf("%s: %w", op, err)
+		}
+		optWrapper := w
+		opts = append(opts, WithWrapper(optWrapper))
```

Comment:
Given that this kind of processing might be performance critical, it's probably better to do a `make([]Option, 0, len(4))` above than to call `append` repeatedly on an initially zero-length slice. Just to avoid possible reallocation.

Reply:
ty!